File: frontend/src/services/leaveService.js
import api from './api';

export async function getAdminLeaves() {
  const { data } = await api.get('/admin/leaves');
  return data;
}

export async function getAdminEmployeeLeaves(employeeId) {
  const { data } = await api.get(`/admin/leaves/employee/${employeeId}`);
  return data;
}

export async function getAdminLeaveById(leaveId) {
  const { data } = await api.get(`/admin/leaves/${leaveId}`);
  return data;
}

export async function approveAdminLeave(leaveId) {
  const { data } = await api.patch(`/admin/leaves/${leaveId}/approve`);
  return data;
}

export async function rejectAdminLeave(leaveId) {
  const { data } = await api.patch(`/admin/leaves/${leaveId}/reject`);
  return data;
}

export async function getManagerLeaves() {
  const { data } = await api.get('/manager/leaves');
  return data;
}

export async function approveManagerLeave(leaveId) {
  const { data } = await api.patch(`/manager/leaves/${leaveId}/approve`);
  return data;
}

export async function rejectManagerLeave(leaveId, reason) {
  const { data } = await api.patch(`/manager/leaves/${leaveId}/reject`, reason ? { reason } : {});
  return data;
}

export async function applyEmployeeLeave(payload) {
  const { data } = await api.post('/employee/leaves', payload);
  return data;
}

export async function getEmployeeLeaves() {
  const { data } = await api.get('/employee/leaves');
  return data;
}

export async function cancelEmployeeLeave(leaveId) {
  const { data } = await api.patch(`/employee/leaves/${leaveId}/cancel`);
  return data;
}

File: backend/src/main/java/com/ems/service/LeaveService.java
package com.ems.service;

import com.ems.dto.request.CreateLeaveRequest;
import com.ems.dto.request.RejectLeaveRequest;
import com.ems.dto.response.LeaveResponse;

import java.util.List;

public interface LeaveService {
    
    // Employee
    LeaveResponse applyLeave(String username, CreateLeaveRequest request);
    List<LeaveResponse> getMyLeaves(String username);
    LeaveResponse getMyLeave(String username, Long leaveId);
    LeaveResponse cancelLeave(String username, Long leaveId);

    // Manager
    List<LeaveResponse> getManagerLeaves(String managerUsername);
    LeaveResponse getManagerLeave(String managerUsername, Long leaveId);
    LeaveResponse approveManagerLeave(String managerUsername, Long leaveId);
    LeaveResponse rejectManagerLeave(String managerUsername, Long leaveId, RejectLeaveRequest request);

    // Admin
    List<LeaveResponse> getAllLeaves();
    List<LeaveResponse> getEmployeeLeaves(Long employeeId);
    LeaveResponse getLeave(Long leaveId);
    LeaveResponse approveAdminLeave(Long leaveId);
    LeaveResponse rejectAdminLeave(Long leaveId);
}
